062-quotas-management.md

```diff
@@ -0,0 +1,163 @@
+2. The initial releases of the Strimzi Quotas Plugin were very simple.
+   It helps to prevent users producing directly to a broker that was running out of disk space.
+   But it fails to prevent replication of messages produced to other brokers in the Kafka cluster from filling in the disk.
+   It also doesn't support many additional features such as JBOD storage.
+   Because the initial releases were very simple, they had also very simple configuration.
+   And it was possible to configure them in the `.spec.kafka.config` section of the `Kafka` CR.
+   The next - not yet released - version of the Strimzi Quotas plugin brings many improvements.
+   More details about how the new version works can be found in the [Cluster Wide Volume Usage Quota Management](./047-cluster-wide-volume-usage-quota-management.md) proposal.
+   It operates at the level of a whole Kafka cluster and it has improved ability to prevent replicated messages to fill in the broker storage.
+   It also supports JBOD storage.
+   But it has also much more complicated configuration.
+   The plugin instances running in the different Kafka brokers communicate with each other using a Kafka topic.
+   And need therefore to connect to the Kafka brokers as clients as well.
+   Having the user configure this directly in the `.spec.kafka.config` section would be much more complicated as the user would need to configure the keystores and truststores, their passswords, bootstrap address etc.
+   Having a dedicated API for quota configuration would make it much easier to have the operator configure the client automatically.
+
+## Proposal
+
+This proposal suggests to add a new `quotas` API to the `Kafka` custom resource.
+It will be inside the `.spec.kafka` section.
+And it will allow configuring different types of quotas.
+As many existing Strimzi APIs, it will use the `type` field to differentiate the quota plugin that will be used.
+
+This proposal suggests to initially add support for two Quota plugins.
+1. The build-in Kafka plugin
+2. The Strimzi Quotas plugin
+
+The first supported plugin - the Kafka's built-in plugin - will be configured as `type: kafka`.
+The API will allow to configure the default user quotas for this plugin in the same format as the `KafkaUser` API allows it today:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: kafka
+      producerByteRate: 1048576
+      consumerByteRate: 2097152
+      requestPercentage: 55
+      controllerMutationRate: 50
+    # ... 
+  # ...
+```
+
+Based on this configuration, the operator will:
+* Configure the default quota plugin (make sure the `client.quota.callback.class` option is not set)
+* Set the default user quotas
+* When the `type: kafka` quotas APi will be set, but one or more of fields with the individual quotas will be missing, the operator will make sure that given default quota is not set.
```

Comment:
By this do you mean the operator will explicitly set empty values for the missing quota types? (and thus the broker inherits from the default configuration) Or jus that it will not set them to some default value?

Reply:
Yes, it is declarative. So not being set in the Kafka CR means it should not be set in the Kafka cluster either. (in this case, the default value for the default quota is null / not set).

Reply:
Ok, that makes sense to me. 

I consider this resolved but don't see a button to mark it as resolved 🤷‍♂️